Prepare internal API for shared use between PB and WM callbacks [1/3]

hmmr opened the pull request:
This is Part 1/3 of #1382 broken up and reassembled into new commits.

After the merge of TTB branch, and in anticipation of the merge of HTTP API branch, this PR proposes a code reorg consisting of the following steps:
1. Move functions shared between PB and WM callbacks from riak_kv_ts_svc to riak_kv_ts_util;
2. Further move functions `get_data`, `put_data`, `delete_data`, `query` and `compile_to_per_quantum_queries` (used to produce coverage results, with `riak_kv_ts_util:sql_to_cover`) into the newly created module `riak_kv_ts_api`. These will constitute the **TS internal API**.
3. Move code serving the INSERT query from `riak_kv_ts_svc` into `riak_kv_qry`. Now the call to `riak_kv_qry:submit("INSERT ...")` will actually take effect (previously, the 'submit' part was a no-op while all the validations, conversions and the eventual call of `riak_kv_ts_util:put_data` happened as a side effect).

Commit:
move exported types and internal defines from .hrl back to .erl

Commit:
move INSERT query code from riak_kv_ts_svc to riak_kv_qry

Commit:
use riak_kv_ts_api funs to figure API call and get perm tokens

with tweaks to decode_query_permissions return value per dialyzer hints.

Commit:
improve error reporting in riak_ts API

* use provided human descriptions for compiler errors instead of forcing
  {MnemonicAtom, Description} through "~p" into #rpberrresp.errmsg;
* handle lexer errors (they are signalled via {'EXIT', {Msg, Stacktrace}},
  so make sure ony Msg is reported to the client.

Commit:
ts_svc: rename for uniformity various make_blah_resp helpers

Commit:
fix validate_row move

Commit:
don't end error messages with a period

Commit:
shorten lines, add spaces around operators

Commit:
move shared functions from ts_svc to ts_util, ts_api

Commit:
riak_ql_ddl:simple_field_type() is now the sole field type

after the maps and other not-simple field types were removed from riak_ql.

hazen commented on src/riak_kv_qry.erl:
```diff
@@ -118,8 +257,10 @@ maybe_await_query_results(_) ->
     receive
-        Result ->
-            Result
+        {ok, Result} ->
```
You are simply guarding against bad results? For proper responses, this should be the same.

hmmr replied:
It's dialyzer wisperings. Because this function returns whatever it receives, dialyzer has no way of knowing what the return type would be. By constraining the terms on reception to only be one of `{ok, Result}` or `{error, Reason}`, we give dialyzer a useful hint.

hazen commented on src/riak_kv_ts_api.erl:
```diff
@@ -0,0 +1,311 @@
+%% -------------------------------------------------------------------
+%%
+%% riak_kv_ts_util: supporting functions for timeseries code paths
```
Should match the name of the module

Commit:
always strip columns info when Rows=[] in make_tsqueryresp

This ensures that both TTB and PB versions of riakc_ts:query get the same
{ok, {[], []}} on empty result set.

Commit:
correct module name in riak_kv_ts_api.erl header

hazen commented on src/riak_kv_ts_error_msgs.hrl:
```diff
@@ -87,3 +87,9 @@
 ).
+
+-define(
+   E_TOO_MANY_SUBQUERIES(N),
```
Why have this as a macro instead of a function? Looks like it's only called in one place anyway.

hmmr replied:
I'm following the local conventions here (alternatively, following the path of least resistance).

hazen commented on src/riak_kv_ts_svc.erl:
```diff
@@ -804,132 +581,130 @@ make_rpberrresp(Code, Message) ->
-    [Mod:get_field_type([N]) || N <- ColumnNames].
 
+make_decoder_error_response({lexer_error, Msg}) ->
+    make_rpberrresp(?E_PARSE_ERROR, flat_format("~s", [Msg]));
```
Did we want these string to be binaries?

hmmr replied:
They all become binaries in `make_rpberrresp/1`. There is no escape.

hazen commented on src/riak_kv_ts_svc.erl:
```diff
@@ -804,132 +581,130 @@ make_rpberrresp(Code, Message) ->
     {tsqueryresp, {[], [], []}};
-make_tsqueryresp({ColumnNames, ColumnTypes, Rows}) ->
-    {tsqueryresp, {ColumnNames, ColumnTypes, Rows}}.
+make_tsqueryresp(Data = {_ColumnNames, _ColumnTypes, _Rows}) ->
```
@hmmr Is this dialyzer again?  I see above that `make_tsgetresp` does almost the same thing, but in a simpler format.

hmmr replied:
The process of producing `ColumnNames` and `ColumnTypes` is different between `riak_kv_ts_api:query` and `:get_data`. For the not-found case to have uniform `{[], []}` presented to the client, columns info needs to be specially stripped from the query return values, while get_data already returns none. Hence this difference in `make_ts{query,get}resp`.

hazen commented on src/riak_kv_ts_api.erl:
```diff
@@ -0,0 +1,311 @@
+
+-spec put_data([[riak_pb_ts_codec:ldbvalue()]], binary()) ->
+                      ok | {error, {some_failed, integer()}} | {error, term()}.
+put_data(Data, Table) ->
```
Nice standardization of get/put operations

hazen commented on src/riak_kv_ts_svc.erl:
```diff
@@ -26,55 +26,87 @@
+        {QryType, SQL} when QryType /= error,
+                            QryType /= 'EXIT' ->
+            {QryType, riak_kv_ts_util:build_sql_record(QryType, SQL, Cover)};
+        {'EXIT', {Reason, _StackTrace}} ->
```
Glad we are catching these now

andytill replied:
If `{QryType, SQL}` was the last pattern then it wouldn't need the guards.

hmmr replied:
You are correct. It would separate the 'good' branches from the ones catching error, and that argument, feeble as it is, has won.

hazen commented on src/riak_kv_w1c_worker.erl:
```diff
@@ -80,27 +80,31 @@ start_link(Name) ->
-                riak_core_apl:get_primary_apl(DocIdx, NVal, riak_kv)
-        end,
+    case riak_core_bucket:get_bucket(Bucket) of
+        {error, Reason} ->
```
Again, glad this error case is being handled

hazen commented:
@hmmr Looks like these are required, too, right?
- https://github.com/basho/riak_pb/pull/190
- https://github.com/basho/riak-erlang-client/pull/277
- https://github.com/basho/riak_test/pull/1047

hmmr commented:
@javajolt All eventually required, yes. In particular, https://github.com/basho/riak_pb/pull/190 is required for #1397, and so also is https://github.com/basho/riak-erlang-client/pull/277. The r_t one uses emended error messages, and is therefore needed for #1396.

My recommendation is to merge the riak_pb PR first, then erlang-client, then proceed with this one.

javajolt commented:
Tests are mostly good.  Investigating failures:
```
Test Results:
ts_simple_activate_table_pass_2-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                       : pass
ts_simple_aggregation-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                 : pass
ts_simple_aggregation_fail-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                            : pass
ts_simple_aggregation_math-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                            : fail
ts_simple_api-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                         : pass
ts_simple_create_table_already_created-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                : pass
ts_simple_create_table_dup_primary_key-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                : pass
ts_simple_create_table_no_primary_key-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                 : pass
ts_simple_create_table_not_null_pk_fields-riak_kv_bitcask_backend,riak_kv_eleveldb_backend             : pass
ts_simple_create_table_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                         : pass
ts_simple_create_table_split_key-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                      : pass
ts_simple_describe_table-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                              : pass
ts_simple_div_by_zero-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                 : fail
ts_simple_get-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                         : pass
ts_simple_get_not_found-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                               : pass
ts_simple_insert-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                      : pass
ts_simple_insert_incorrect_columns-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                    : pass
ts_simple_latin1_create_table_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend             : pass
ts_simple_put-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                         : pass
ts_simple_put_all_datatypes-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                           : pass
ts_simple_put_all_null_datatypes-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                      : pass
ts_simple_put_bad_date-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                : pass
ts_simple_put_invalid_data-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                            : pass
ts_simple_put_non_existent_bucket-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                     : pass
ts_simple_recompile_ddl-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                               : pass
ts_simple_select-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                      : pass
ts_simple_select_compare_two_fields_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend       : pass
ts_simple_select_double_in_key-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                        : pass
ts_simple_select_incompatible_type_float_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend  : pass
ts_simple_select_incompatible_type_integer_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend: pass
ts_simple_select_missing_field_in_pk_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend      : pass
ts_simple_select_not_found-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                            : pass
ts_simple_select_unexpected_token_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend         : pass
ts_simple_select_where_has_no_lower_bounds_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend: pass
ts_simple_select_where_has_no_upper_bounds_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend: pass
ts_simple_single_key_ops-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                              : pass
ts_simple_unicode_create_table_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend            : pass
riak_shell_test_connecting-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                            : pass
riak_shell_test_connecting_error-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                      : pass
riak_shell_test_disconnecting-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                         : pass
riak_shell_test_reconnecting-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                          : pass
ts_cluster_activate_table_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                      : pass
ts_cluster_aggregation-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                : pass
ts_cluster_comprehensive-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                              : fail
ts_cluster_coverage-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                   : pass
ts_cluster_create_table_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                        : pass
ts_cluster_create_table_via_sql_SUITE-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                 : pass
ts_cluster_keys_SUITE-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                 : pass
ts_cluster_put_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                 : pass
ts_cluster_quantum_boundaries_SUITE-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                   : pass
ts_cluster_riak_shell_basic_sql-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                       : pass
ts_cluster_riak_shell_regression_log-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                  : fail
ts_cluster_select_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                              : pass
ts_cluster_select_pass_2-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                              : pass
ts_cluster_select_pass_3_sorted_on_key-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                : pass
ts_cluster_unicode-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                    : fail
ts_degraded_activate_table_fail_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                     : pass
ts_degraded_aggregation-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                               : pass
ts_degraded_select_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                             : pass
ts_degraded_select_pass_2-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                             : pass
ts_simple_activate_table_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                       : pass
---------------------------------------------
5 Tests Failed
56 Tests Passed
That's 91.80327868852459% for those keeping score
```

hazen commented:
Here are explanation of failures:
- [ ]  `ts_simple_aggregation_math/ts_simple_div_by_zero` - Got back `divide_by_zero` instead of `Divide by Zero`
- [x] `ts_cluster_comprehensive/ts_unicode` - Result from creating table now needs to be `{ok, {_,_}}`
- [ ] `ts_cluster_riak_shell_regression_log` - Got back `too_many_subqueries` but expected `Too many subqueries (7)`

hazen commented:
```
Test Results:
ts_simple_activate_table_pass_2-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                       : pass
ts_simple_aggregation-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                 : pass
ts_simple_aggregation_fail-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                            : pass
ts_simple_aggregation_math-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                            : pass
ts_simple_api-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                         : pass
ts_simple_create_table_already_created-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                : pass
ts_simple_create_table_dup_primary_key-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                : pass
ts_simple_create_table_no_primary_key-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                 : pass
ts_simple_create_table_not_null_pk_fields-riak_kv_bitcask_backend,riak_kv_eleveldb_backend             : pass
ts_simple_create_table_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                         : pass
ts_simple_create_table_split_key-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                      : pass
ts_simple_describe_table-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                              : fail
ts_simple_div_by_zero-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                 : pass
ts_simple_get-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                         : pass
ts_simple_get_not_found-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                               : pass
ts_simple_insert-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                      : pass
ts_simple_insert_incorrect_columns-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                    : pass
ts_simple_latin1_create_table_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend             : pass
ts_simple_put-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                         : pass
ts_simple_put_all_datatypes-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                           : pass
ts_simple_put_all_null_datatypes-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                      : pass
ts_simple_put_bad_date-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                : pass
ts_simple_put_invalid_data-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                            : pass
ts_simple_put_non_existent_bucket-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                     : pass
ts_simple_recompile_ddl-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                               : pass
ts_simple_select-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                      : pass
ts_simple_select_compare_two_fields_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend       : pass
ts_simple_select_double_in_key-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                        : pass
ts_simple_select_incompatible_type_float_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend  : pass
ts_simple_select_incompatible_type_integer_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend: pass
ts_simple_select_missing_field_in_pk_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend      : pass
ts_simple_select_not_found-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                            : pass
ts_simple_select_unexpected_token_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend         : pass
ts_simple_select_where_has_no_lower_bounds_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend: pass
ts_simple_select_where_has_no_upper_bounds_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend: pass
ts_simple_single_key_ops-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                              : pass
ts_simple_unicode_create_table_not_allowed-riak_kv_bitcask_backend,riak_kv_eleveldb_backend            : pass
riak_shell_test_connecting-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                            : pass
riak_shell_test_connecting_error-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                      : pass
riak_shell_test_disconnecting-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                         : pass
riak_shell_test_reconnecting-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                          : pass
ts_cluster_activate_table_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                      : pass
ts_cluster_aggregation-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                : pass
ts_cluster_comprehensive-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                              : pass
ts_cluster_coverage-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                   : fail
ts_cluster_create_table_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                        : pass
ts_cluster_create_table_via_sql_SUITE-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                 : fail
ts_cluster_keys_SUITE-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                 : pass
ts_cluster_put_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                 : pass
ts_cluster_quantum_boundaries_SUITE-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                   : pass
ts_cluster_riak_shell_basic_sql-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                       : fail
ts_cluster_riak_shell_regression_log-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                  : fail
ts_cluster_select_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                              : pass
ts_cluster_select_pass_2-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                              : pass
ts_cluster_select_pass_3_sorted_on_key-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                : pass
ts_cluster_unicode-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                                    : pass
ts_degraded_activate_table_fail_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                     : pass
ts_degraded_aggregation-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                               : pass
ts_degraded_select_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                             : pass
ts_degraded_select_pass_2-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                             : pass
ts_simple_activate_table_pass_1-riak_kv_bitcask_backend,riak_kv_eleveldb_backend                       : pass
---------------------------------------------
5 Tests Failed
56 Tests Passed
That's 91.80327868852459% for those keeping score
```

hazen commented:
- [ ] `ts_simple_describe_table` is returning a list of lists instead of a list of tuples for data
- [ ] `ts_cluster_coverage` function clause problem calling `riak_kv_ts_svc:compile` in `ts_cluster_coverage:test_quanta_range`
- [ ] `ts_cluster_create_table_via_sql_SUITE` fails in `describe_test`. Probably same as above.
- [ ] `ts_cluster_riak_shell_basic_sql` same describe failure
- [ ] `ts_cluster_riak_shell_regression_log-riak_kv_bitcask_backend` got `{{badmatch,{error,{already_started,riak_shell}}},[{riak_shell_app,boot_TEST,1,[{file,"src/riak_shell_app.erl"},{line,43}]},{riak_shell_test_util,shell_init,0,[{file,"tests/riak_shell_test_util.erl"},{line,49}]}`. Guessing previous test failure left the shell running. Unrelated to this PR.

Commit:
riak_kv_qry:do_describe/1 now returns a list of tuples

Commit:
Fix results of decode_query/2 in sub_tscoveragereq/3

Commit:
Fix unit test and silence dialyzer

hazen commented:
👍 169492a

hmmr commented:
What's the "seventy" code for _Amen_?
